File: Prefix Sum/Python/Largest subarray of 0's and 1's.py
class Solution:
    def maxLen(self, arr):
        # code here
        mp={}
        preSum=0
        res=0
        
        for i in range(len(arr)):
            preSum+=-1 if arr[i]==0 else 1
            
            if preSum==0:
                res=i+1
            
            if preSum in mp:
                res=max(res,i-mp[preSum])
            else:
                mp[preSum]=i
        return res

File: Prefix Sum/C++/Largest subarray of 0's and 1's.cpp
class Solution {
	public:
	int maxLen(vector<int> &arr) {
		// Your code here
		unordered_map<int, int> mp;
		int res = 0;
		int preSum = 0;
		
		for (int i = 0; i< arr.size(); i++) {
		    preSum += (arr[i] == 0) ? -1 : 1;
		    
		    if (preSum == 0)
		        res = i + 1;
		    if (mp.find(preSum) != mp.end())
		        res = max(res, i - mp[preSum]);
		    else
		        mp[preSum] = i;
		}
		return res;
	}
};
